### PR TITLE
add ForEachKey and ForEachItem methods

### PR DESCRIPTION
**Pull request by mingpen**
[add ForEachKey and ForEachItem methods](https://github.com/DeanThompson/syncmap/issues/1)

### CODE REVIEW

**Comment by DeanThompson** [comment]
Hi @mingpen , thanks for the contribution. Some thoughts here:

1. Please run `go fmt` before committing.
2. The `ErrorStop` is useless in this package, and can be removed.
3. I suggest we should implement two kinds of callback functions:

a) `func(key string)` or `func(item *Item)`, these callbacks simply receive each key or item, and loop over until no more keys or items. There is no need to care about errors.

b) `func(key string) bool` or `func(item *Item) bool`. These callbacks provide the possibility to break out of the loop by returning `false`.

**Comment by mingpen** [comment]
there is a new commit, expect @DeanThompson your check.  
thanks!

File: syncmap.go [hunk 1]
```diff
@@ -133,17 +133,51 @@ func (m *SyncMap) Flush() int {
 	return size
 }
 
+// IterKeyFun is the type of the function called for each key.
+//
+// If false is returned,each key stops.
+// Don't modify the SyncMap in this function,
+// or maybe leads to deadlock.
+type IterKeyFun func(key string) bool
```
Comment (DeanThompson) [inline syncmap.go]: Rename the type to `IterKeyWithBreakFunc`, or just remove this type, and change the method signature to

```go
func (m *SyncMap) EachKeyWithBreak(f func(key string) bool)
```

File: syncmap.go [hunk 2]
```diff
@@ -133,17 +133,51 @@ func (m *SyncMap) Flush() int {
...
+
+func (m *SyncMap) ForEachKey(iter IterKeyFun) {
```
Comment (DeanThompson) [inline syncmap.go]: Rename this method to `EachKeyWithBreak`

File: syncmap.go [hunk 3]
```diff
@@ -133,17 +133,51 @@ func (m *SyncMap) Flush() int {
...
+	bStop := false
```
Comment (DeanThompson) [inline syncmap.go]: What is `bStop` short for? How about renaming to `shouldStop` or simply `stop`?

File: syncmap.go [hunk 4]
```diff
@@ -133,17 +133,51 @@ func (m *SyncMap) Flush() int {
...
+	}
+}
+
+// IterEveryKeyFun is the type of the function called for every key.
+//
+// Don't modify the SyncMap in this function,
+// or maybe leads to deadlock.
+type IterEveryKeyFun func(key string)
+
+func (m *SyncMap) IterEveryKey(iter IterEveryKeyFun) {
```
Comment (DeanThompson) [inline syncmap.go]: Rename this method to `EachKey`

File: syncmap.go [hunk 5]
```diff
@@ -133,17 +133,51 @@ func (m *SyncMap) Flush() int {
...
+			break
+		}
+	}
+}
+
+// IterEveryKeyFun is the type of the function called for every key.
+//
+// Don't modify the SyncMap in this function,
+// or maybe leads to deadlock.
+type IterEveryKeyFun func(key string)
```
Comment (DeanThompson) [inline syncmap.go]: Rename the type to `IterKeyFunc`, or just remove this type, and change the method signature to

```go
func (m *SyncMap) EachKey(f func(key string))
```

File: syncmap.go [hunk 6]
```diff
@@ -155,17 +189,52 @@ type Item struct {
 	Value interface{}
 }
 
+// IterItemFun is the type of the function called for each item.
+//
+// If false is returned,each item stops.
+// Don't modify the SyncMap in this function,
+// or maybe leads to deadlock.
+type IterItemFun func(item *Item) bool
```
Comment (DeanThompson) [inline syncmap.go]: Rename the type to `IterItemWithBreakFunc`, or just remove this type, and change the method signature to

```go
func (m *SyncMap) EachItemWithBreak(f func(item *Item) bool)
```

File: syncmap.go [hunk 7]
```diff
@@ -155,17 +189,52 @@ type Item struct {
...
+
+func (m *SyncMap) ForEachItem(iter IterItemFun) {
```
Comment (DeanThompson) [inline syncmap.go]: Rename this method to `EachItemWithBreak`

File: syncmap.go [hunk 8]
```diff
@@ -155,17 +189,52 @@ type Item struct {
...
+		}
+	}
+
+}
+
+// IterEveryItemFun is the type of the function called for every item.
+//
+// Don't modify the SyncMap in this function,
+// or maybe leads to deadlock.
+type IterEveryItemFun func(item *Item)
```
Comment (DeanThompson) [inline syncmap.go]: Rename the type to `IterItemFunc`, or just remove this type, and change the method signature to

```go
func (m *SyncMap) EachItem(f func(item *Item))
```

File: syncmap.go [hunk 9]
```diff
@@ -155,17 +189,52 @@ type Item struct {
...
+
+func (m *SyncMap) IterEveryItem(iter IterEveryItemFun) {
```
Comment (DeanThompson) [inline syncmap.go]: Rename this method to `EachItem`

**Review by DeanThompson — COMMENTED** [review]
Hi @mingpen , thanks again!

I wrote some comments that basically are about naming pattern. Please take a look and push one more commit so that we can merge to the master branch.

File: syncmap.go (hunk 3 above)
Comment (mingpen) [reply syncmap.go]: type(bool) + name
stop is ok

**Comment by mingpen** [comment]
I pushed one commit.  
Please take a look ^_^ @DeanThompson

**Review by DeanThompson — APPROVED** [review]
Thanks for the effort! Merged to master.

**Pull request merged**